xen/arch/arm/cpufreq/arm_scpi.c

```diff
@@ -566,21 +566,38 @@ static int scpi_send_message(u8 idx, void *tx_buf, unsigned int tx_len,
 	reinit_completion(&msg->done);
 
 	ret = mbox_send_message(scpi_chan->chan, msg);
-	if (ret < 0 || !rx_buf)
+	if (ret < 0 || !rx_buf) {
+		/* mbox_send_message returns non-negative value on success */
+		ret2 = ret < 0 ? ret : 0;
 		goto out;
+	}
 
 	if (!wait_for_completion_timeout(&msg->done, MAX_RX_TIMEOUT))
 		ret = -ETIMEDOUT;
 	else
 		/* first status word */
 		ret = msg->status;
+
+	/* SCPI error codes > 0, translate them to Linux scale */
+	ret2 = ret > 0 ? scpi_to_linux_errno(ret) : ret;
+
+	/*
+	 * If the mailbox controller has TX-done irq it definitely knows when
+	 * transmitted data has been sent and it is responsible for calling
+	 * mbox_chan_txdone() to signal framework about TX-done.
+	 * If controller can't generate TX-Done irq the client has to signal
+	 * TX-done by itself. It is safe to call mbox_client_txdone() after
+	 * receiving a response or getting a timeout error.
```

Comment:
> It is safe to call mbox_client_txdone() after ... getting a timeout error.

This is somehow true in our particular case. Because we know that in case of asynchronous MB we will always have a timeout. And pray timeout is big enough to let other end have done its job. But that statement is uneven in a generic case.